cmd/docsync/main.go

```diff
@@ -0,0 +1,169 @@
+// files to upload.
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"mime"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+func main() {
+	bucket := os.Args[1]
+	bucketPrefix := os.Args[2]
+	localDir := os.Args[3]
+
+	localKeys := mustListContents(localDir)
+
+	region := "us-east-1"
+	sess := session.Must(session.NewSession(aws.NewConfig().WithRegion(region)))
+	svc := s3.New(sess)
+
+	var remoteKeys []string
+	err := svc.ListObjectsPages(&s3.ListObjectsInput{
+		Bucket: aws.String(bucket),
+		Prefix: aws.String(bucketPrefix),
+	}, func(page *s3.ListObjectsOutput, lastPage bool) bool {
+		for _, obj := range page.Contents {
+			if !(*obj.Key == "docs/") {
+				remoteKeys = append(remoteKeys, *obj.Key)
+			}
+		}
+		return true
+	})
+
+	if err != nil {
+		log.Fatalln("s3 list objects error:", err)
+	}
+
+	for _, k := range localKeys {
+		var body []byte
+
+		path := fmt.Sprintf("%s/%s", localDir, k)
+		body, err = ioutil.ReadFile(path)
+		if err != nil {
+			log.Fatalln(err.Error())
+		}
+
+		upload := &s3.PutObjectInput{
+			Bucket: aws.String(bucket),
+			Key:    aws.String(bucketPrefix + k),
+			Body:   bytes.NewReader(body),
+		}
+
+		ext := filepath.Ext(path)
+		contentType := mime.TypeByExtension(ext)
+
+		if contentType == "" {
+			upload.SetContentType("text/html")
+		} else {
+			upload.SetContentType(contentType)
+		}
+
+		_, err = svc.PutObject(upload)
+
+		if err != nil {
+			log.Fatalln(err.Error())
+		}
+	}
+
+	var prefixedLocalKeys []string
+	for _, k := range localKeys {
+		prefixedLocalKeys = append(prefixedLocalKeys, path.Join("docs", k))
+	}
+
+	remoteOnly := setDiff(remoteKeys, prefixedLocalKeys)
+	for _, k := range remoteOnly {
+		remove := &s3.DeleteObjectInput{
+			Bucket: aws.String(bucket),
+			Key:    aws.String(k),
+		}
+
+		_, err = svc.DeleteObject(remove)
+
+		if err != nil {
+			log.Fatalln(err.Error())
+		}
+	}
+}
+
+func setDiff(a, b []string) []string {
+	// don't modify input
+	a = append([]string{}, a...)
+	b = append([]string{}, b...)
+	sort.Strings(a)
+	sort.Strings(b)
+
+	var (
+		diff []string
+		i, j int
+	)
+
+	for {
+		if i == len(a) {
+			break
+		}
+
+		if j == len(b) {
+			diff = append(diff, a[i:]...)
+			break
+		}
+
+		if a[i] < b[j] {
+			diff = append(diff, a[i])
+			i++
+		} else if b[j] < a[i] {
+			j++
+		} else { // a[i] == b[j]
+			i++
+			j++
+		}
+	}
+
+	return diff
+}
+
+func mustListContents(parentPath string) []string {
+	files, err := ioutil.ReadDir(parentPath)
+	if err != nil {
+		log.Fatalln("ReadDir error:", err)
+	}
+
+	var res []string
+	for _, f := range files {
+		n := f.Name()
+
+		if f.IsDir() {
+			descendants := mustListContents(path.Join(parentPath, n))
+			for _, d := range descendants {
+				res = append(res, path.Join(n, d))
+			}
+		} else {
+			if !strings.HasPrefix(n, ".") {
```

Comment:
We should probably skip dot directories too, right? Maybe move this out of the `if/else` section here and do this instead:

```
if string.HasPrefix(n) {
  continue
}
```